compiler/scalc/ast.ml

```diff
@@ -37,41 +37,71 @@ let dead_value = VarName.fresh ("dead_value", Pos.no_pos)
 let handle_default = FuncName.fresh ("handle_default", Pos.no_pos)
 let handle_default_opt = FuncName.fresh ("handle_default_opt", Pos.no_pos)
 
-type operator =
-  < overloaded : no ; monomorphic : yes ; polymorphic : yes ; resolved : yes >
-  Shared_ast.operator
+type operator = Shared_ast.lcalc Shared_ast.operator
 
 type expr = naked_expr Mark.pos
 
 and naked_expr =
-  | EVar : VarName.t -> naked_expr
-  | EFunc : FuncName.t -> naked_expr
-  | EStruct : expr list * StructName.t -> naked_expr
-  | EStructFieldAccess : expr * StructField.t * StructName.t -> naked_expr
-  | EInj : expr * EnumConstructor.t * EnumName.t -> naked_expr
-  | EArray : expr list -> naked_expr
-  | ELit : lit -> naked_expr
-  | EApp : expr * expr list -> naked_expr
-  | EAppOp : operator * expr list -> naked_expr
+  | EVar of VarName.t
+  | EFunc of FuncName.t
+  | EStruct of { fields : expr StructField.Map.t; name : StructName.t }
+  | EStructFieldAccess of {
+      e1 : expr;
```

Comment:
Why `e1` and not just `e` since there is only one ?

Reply:
Fear of variable shadowing in pattern matching...

Reply:
I don't think syntactic sugar (punning) should direct struct field namings... but I'm nitpicking here, I don't really mind, I was mostly curious.